File: algorithms/dynamic-programming/knuth-optimization/kotlin/KnuthOptimization.kt
fun knuthOptimization(n: Int, freq: IntArray): Int {
    if (n == 0) return 0
    val dp = Array(n) { IntArray(n) }
    val opt = Array(n) { IntArray(n) }
    val prefix = IntArray(n + 1)
    for (i in 0 until n) prefix[i + 1] = prefix[i] + freq[i]

    for (i in 0 until n) {
        dp[i][i] = freq[i]
        opt[i][i] = i
    }

    for (len in 2..n) {
        for (i in 0..n - len) {
            val j = i + len - 1
            dp[i][j] = Int.MAX_VALUE
            val costSum = prefix[j + 1] - prefix[i]
            val lo = opt[i][j - 1]
            val hi = if (i + 1 <= j) opt[i + 1][j] else j
            for (k in lo..hi) {
                val left = if (k > i) dp[i][k - 1] else 0
                val right = if (k < j) dp[k + 1][j] else 0
                val v = left + right + costSum
                if (v < dp[i][j]) {
                    dp[i][j] = v
                    opt[i][j] = k
                }
            }
        }
    }
    return dp[0][n - 1]
}

fun main() {
    val input = System.`in`.bufferedReader().readText().trim().split("\\s+".toRegex()).map { it.toInt() }
    val n = input[0]
    val freq = input.subList(1, 1 + n).toIntArray()
    println(knuthOptimization(n, freq))
}

File: algorithms/dynamic-programming/knuth-optimization/typescript/knuthOptimization.ts
export function knuthOptimization(n: number, freq: number[]): number {
    const dp: number[][] = Array.from({ length: n }, () => new Array(n).fill(0));
    const opt: number[][] = Array.from({ length: n }, () => new Array(n).fill(0));
    const prefix: number[] = new Array(n + 1).fill(0);
    for (let i = 0; i < n; i++) prefix[i + 1] = prefix[i] + freq[i];

    for (let i = 0; i < n; i++) {
        dp[i][i] = freq[i];
        opt[i][i] = i;
    }

    for (let len = 2; len <= n; len++) {
        for (let i = 0; i <= n - len; i++) {
            const j = i + len - 1;
            dp[i][j] = Number.MAX_SAFE_INTEGER;
            const costSum = prefix[j + 1] - prefix[i];
            const lo = opt[i][j - 1];
            const hi = i + 1 <= j ? opt[i + 1][j] : j;
            for (let k = lo; k <= hi; k++) {
                const left = k > i ? dp[i][k - 1] : 0;
                const right = k < j ? dp[k + 1][j] : 0;
                const val = left + right + costSum;
                if (val < dp[i][j]) {
                    dp[i][j] = val;
                    opt[i][j] = k;
                }
            }
        }
    }
    return dp[0][n - 1];
}
